src/loss.jl

```diff
@@ -101,6 +85,56 @@ mutable structdef enum
 
 =#          
 
+
+"""
+    softmax(x;[dims])
+
+The softmax function typically used in classification.   
+Gives the same results as to `exp.(logp(x, dims))`.   
+See also `logsoftmax`.   
+"""
+softmax(x;dims=:,algo=1)=generic_softmax(x,algo,_softmax;dims=dims)
+
+function _softmax(x; dims=:)
+    x = x .- maximum(x;dims=dims)
+    x = exp.(x)
+    return x ./ sum(x;dims=dims)
+end
+
+function _softback(x,y,dy;dims=:)
+    return y .* dy .- y .* sum(y .* dy; dims=dims)
+end
+
+@primitive  _softmax(x;dims=:),dy,y  _softback(x,y,dy,dims=dims)
+
+"""
+     logsoftmax(x;[dims])
+
+ Equivalent to `logp(x;[dims])`. See also `sotfmax`. 
+"""
+const logsoftmax = logp
+
+function generic_softmax(x,algo::Int,fallback;dims=:)
+    if isa(value(x), KnetArray)
+        if dims==1
+            sz = size(x)
+            x = cudnnSoftmaxForward(reshape(x, (1,1,sz[1],:)), algo=algo)
+            reshape(x, sz)
+        elseif dims==2 && ndims==2
```

Comment:
ndims -> ndims(x)

An error like this should be caught when tested on a gpu. If not, please add a test

Reply:
tests added, typo is corrected.